docker/android-system.sh

```diff
@@ -46,64 +79,474 @@ EOF
+echo "OpenJDK Runtime Environment (${jre_info})"
+echo "OpenJDK 64-Bit Server VM (${build_info})"
+EOF
+
+    cat << EOF > /usr/bin/javac
+#!/usr/bin/env bash
+echo "javac ${java_version}"
+EOF
+
+    chmod +x /usr/bin/java
+    chmod +x /usr/bin/javac
+
+    # more faking
+    export ANDROID_JAVA_HOME=/tmp
+    mkdir -p /tmp/lib/
+    touch /tmp/lib/tools.jar
+}
+
+build_android() {
+    if [[ "${ANDROID_SYSTEM_COMPLETE}" != "1" ]]; then
+        export ALLOW_MISSING_DEPENDENCIES=true
+    fi
+
+    set +u
+    # shellcheck disable=SC1091
+    source build/envsetup.sh
+    lunch "aosp_${ARCH}-user"
+    if [[ "${ANDROID_SYSTEM_COMPLETE}" != "1" ]]; then
+        mmma bionic/
+        mmma external/mksh/
+        mmma system/core/toolbox/
+    else
+        mma
+    fi
+    if [[ "${ANDROID_SYSTEM_COMPLETE}" != "1" ]] && [[ "${MAJOR_VERSION}" -ge 11 ]]; then
+        # for some reason, building bionic doesn't build linker64 on the first pass
+        # doing a partial build and a rebuild is just as fast though.
+        rm -rf out/target/product/generic
+        mmma bionic/
+    fi
+    set -u
+}
+
+install_android() {
+    local outdir=
+    if [[ "${ARCH}" = "arm" ]]; then
+        outdir=out/target/product/generic
+    else
+        outdir="out/target/product/generic_${ARCH}"
+    fi
+    mv "${outdir}/system/" /
+    if [[ "${ANDROID_SYSTEM_COMPLETE}" == "1" ]] && [[ -d "${outdir}/apex" ]]; then
+        # can use the APEX linker, no need to use the bootstrap one
+        mv "${outdir}/apex/" /
+    elif [[ "${MAJOR_VERSION}" -ge 10 ]]; then
+        symlink_bootstrap
+    fi
+
+    # list from https://elinux.org/Android_toolbox
+    local tool=
+    for tool in "${@}"; do
+        if [[ ! -f "/system/bin/${tool}" ]]; then
+            ln -s /system/bin/toolbox "/system/bin/${tool}"
+        fi
+    done
+
+    echo "127.0.0.1 localhost" > /system/etc/hosts
+}
+
+symlink_bootstrap() {
+    # for Android 10+, we need to use the bootstrap linker rather than
+    # the APEX linker, which is gigantic. we also symlink the ASAN
+    # linker just in case using the bootstrapped one.
+    local linker
+    local file
+
+    if compgen -G /system/bin/bootstrap/* >/dev/null 2>&1; then
+        for linker in /system/bin/bootstrap/*; do
+            file=$(basename "${linker}")
+            unlink "/system/bin/${file}"
+            ln -s "/system/bin/bootstrap/${file}" "/system/bin/${file}"
+        done
+    fi
+
+    # also need to ensure the shared libraries aren't symlinks
+    local lib
+    local libdir
+    for libdir in /system/lib{,64}; do
+        if compgen -G "${libdir}/bootstrap/"* >/dev/null 2>&1; then
+            for lib in "${libdir}/bootstrap/"*; do
+                file=$(basename "${lib}")
+                unlink "${libdir}/${file}"
+                ln -s "${libdir}/bootstrap/${file}" "${libdir}/${file}"
+            done
+        fi
+    done
+}
+
+# this are the minimum set of modules that are need to build bionic
```

Comment:
I've separated the logic for each Android version, rather than do a series of if chains, because it makes it much less clear what the specific dependencies for each version are. This simplifies maintenance a lot, in case something breaks for a specific Android version but not another patch or minor version.